### PR TITLE
feat: capture client error in the network interceptor

### PR DESCRIPTION
**Pull request by ahmedAlaaInstabug**
## Description of the change

1) support capturing client error in the network interceptor

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

JIRA ID : MOB-14766

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests

### Code review

- [ ] This pull request has a descriptive title and information useful to a reviewer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -7,6 +7,10 @@
 - Fix APM network logging on Android ([#1253](https://github.com/Instabug/Instabug-React-Native/pull/1253)).
 - Fix an OOM (out-of-memory) crash while saving network logs on Android ([#1244](https://github.com/Instabug/Instabug-React-Native/pull/1244)).
 
+### Added
+
+- Add support for capturing client error message in network interceptor([#1256](https://github.com/Instabug/Instabug-React-Native/pull/1256)).
```
Comment (AndrewAminInstabug) [inline CHANGELOG.md:12]: we can improve this like 

> Enhance the network interceptor to capture error messages from the client.

Comment (AndrewAminInstabug) [inline CHANGELOG.md:12]: we can improve this like so

> Enhance the network interceptor to capture error messages from the client.

**Review by a7medev — CHANGES_REQUESTED** [review]

File: src/utils/XhrNetworkInterceptor.ts [hunk 2]
```diff
@@ -150,9 +150,15 @@ export default {
 
               // @ts-ignore
               const _response = this._response;
-              cloneNetwork.requestBody =
-                typeof _response === 'string' ? _response : JSON.stringify(_response);
-              cloneNetwork.responseBody = null;
+              // detect error message when the make xhr requests like using axios
+              if (_response != null && _response.toString().length > 0) {
+                cloneNetwork.errorDomain = typeof _response === 'string' ? _response : '';
+              }
```
Comment (a7medev) [inline src/utils/XhrNetworkInterceptor.ts:156]: I think this will cause the `errorDomain` to be empty signaling no error when there has actually been an error in the network request, I think we should set `errorDomain` to `_response` if it's a string and if not we should keep it as `ClientError`, what do you think?

```suggestion
              // Detect a more descriptive error message.
              if (typeof _response === 'string' && _response.length > 0) {
                cloneNetwork.errorDomain = _response;
              }
```

**Review by a7medev — CHANGES_REQUESTED** [review]

**Review by a7medev — APPROVED** [review]
🚀

File: CHANGELOG.md [hunk 3]
```diff
@@ -7,6 +7,10 @@
 - Fix APM network logging on Android ([#1253](https://github.com/Instabug/Instabug-React-Native/pull/1253)).
 - Fix an OOM (out-of-memory) crash while saving network logs on Android ([#1244](https://github.com/Instabug/Instabug-React-Native/pull/1244)).
 
+### Added
+
+- Enhance the network interceptor to capture some client error messages ([#1256](https://github.com/Instabug/Instabug-React-Native/pull/1256)).
+
```
Comment (a7medev) [inline CHANGELOG.md:13]: Can you please move the "Added" section before the "Fixed" section as described in the Changelog guidelines? 🙏🏼

**Review by a7medev — APPROVED** [review]

File: CHANGELOG.md [hunk 4]
```diff
@@ -4,6 +4,7 @@
 
 ### Added
 
+- Enhance the network interceptor to capture some client error messages ([#1256](https://github.com/Instabug/Instabug-React-Native/pull/1256)).
```
Comment (a7medev) [inline CHANGELOG.md:7]:
```suggestion
- Enhance the network interceptor to capture some client error messages ([#1257](https://github.com/Instabug/Instabug-React-Native/pull/1257)).
```

File: CHANGELOG.md [hunk 5]
```diff
@@ -79,17 +80,17 @@
...
 
 ### Added
 
 - Adds symbol files upload script ([#1137](https://github.com/Instabug/Instabug-React-Native/pull/1137))
 - Support enabling NDK crash capturing on Android ([#1132](https://github.com/Instabug/Instabug-React-Native/pull/1132)).
 
+### Deprecated
+
+- Deprecate Execution Traces APIs `APM.startExecutionTrace`, `Trace.end` and `Trace.setAttribute` in favor of the new App Flows APIs ([#1138](https://github.com/Instabug/Instabug-React-Native/pull/1138)).
+
```
Comment (a7medev) [inline CHANGELOG.md:93]: I think this change doesn't belong to this PR 👀

File: src/utils/XhrNetworkInterceptor.ts [hunk 6]
```diff
@@ -72,7 +72,8 @@ export default {
     originalXHROpen = XMLHttpRequest.prototype.open;
     originalXHRSend = XMLHttpRequest.prototype.send;
     originalXHRSetRequestHeader = XMLHttpRequest.prototype.setRequestHeader;
-
+    // An error code that signifies an issue with the RN client.
+    const clientErrorCode = 9876;
```
Comment (a7medev) [inline src/utils/XhrNetworkInterceptor.ts:76]: Did we agree on this somewhere? 👀

File: src/utils/XhrNetworkInterceptor.ts [hunk 7]
```diff
@@ -145,14 +146,20 @@ export default {
...
               cloneNetwork.errorDomain = 'ClientError';
 
               // @ts-ignore
               const _response = this._response;
-              cloneNetwork.requestBody =
-                typeof _response === 'string' ? _response : JSON.stringify(_response);
-              cloneNetwork.responseBody = null;
+              // Detect a more descriptive error message.
+              if (typeof _response === 'string' && _response.length > 0) {
+                cloneNetwork.errorDomain = _response.replaceAll(' ', '-');
```
Comment (a7medev) [inline src/utils/XhrNetworkInterceptor.ts:156]: We are still using `-` instead of spaces because of the backend limitation, right? If so, do we have a backlog item for changing this in the future?

**Review by a7medev — CHANGES_REQUESTED** [review]

File: src/utils/XhrNetworkInterceptor.ts (hunk 6 above)
Comment (ahmedAlaaInstabug) [reply src/utils/XhrNetworkInterceptor.ts:76]: yes we agree it with backend team

**Review by a7medev — DISMISSED** [review]

**Review by AyaMahmoud148 — DISMISSED** [review]

File: CHANGELOG.md [hunk 8]
```diff
@@ -5,6 +5,7 @@
 ### Added
 
 - Add support for Expo Router navigation tracking ([#1270](https://github.com/Instabug/Instabug-React-Native/pull/1270)).
+- Enhance the network interceptor to capture some client error messages ([#1257](https://github.com/Instabug/Instabug-React-Native/pull/1257)).
```
Comment (AndrewAminInstabug) [inline CHANGELOG.md:8]: can we improve it ... 
 [Enhance the network interceptor to capture **more** client error messages]

File: examples/default/src/screens/apm/NetworkScreen.tsx [hunk 9]
```diff
@@ -45,6 +46,32 @@ export const NetworkScreen: React.FC = () => {
...
 
+  async function sendRequestToUrlUsingAxios() {
+    let urlToSend = '';
+
+    if (endpointUrl.trim() !== '') {
+      urlToSend = endpointUrl;
+      console.log('Sending request to: ', endpointUrl);
+    } else {
+      // Use json placeholder URL as a default if endpointUrl is empty
+      console.log('sending request to default json placeholder');
```
Comment (AndrewAminInstabug) [inline examples/default/src/screens/apm/NetworkScreen.tsx:57]: QQ: Do we use `console.log()` for production? This does not respect the log levels.

Comment (ahmedAlaaInstabug) [reply examples/default/src/screens/apm/NetworkScreen.tsx:57]: These logs are found in the example app.

**Review by a7medev — APPROVED** [review]

**Review by AndrewAminInstabug — DISMISSED** [review]

**Review by AndrewAminInstabug — APPROVED** [review]

**Review by a7medev — APPROVED** [review]
